processing/src/main/java/org/apache/druid/query/scan/OrderByLimitQueryRunner.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.scan;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Sets;
+import org.apache.druid.collections.MultiColumnSorter;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.java.util.common.JodaUtils;
+import org.apache.druid.java.util.common.granularity.Granularities;
+import org.apache.druid.java.util.common.guava.Sequence;
+import org.apache.druid.java.util.common.guava.Sequences;
+import org.apache.druid.query.Query;
+import org.apache.druid.query.QueryContexts;
+import org.apache.druid.query.QueryMetrics;
+import org.apache.druid.query.QueryPlus;
+import org.apache.druid.query.QueryRunner;
+import org.apache.druid.query.context.ResponseContext;
+import org.apache.druid.query.filter.Filter;
+import org.apache.druid.segment.Cursor;
+import org.apache.druid.segment.Segment;
+import org.apache.druid.segment.StorageAdapter;
+import org.apache.druid.segment.VirtualColumn;
+import org.apache.druid.segment.column.ColumnHolder;
+import org.apache.druid.segment.filter.Filters;
+import org.apache.druid.timeline.SegmentId;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+class OrderByLimitQueryRunner implements QueryRunner<ScanResultValue>
```

Comment:
I made an earlier comment about the complexity that results from combining sorting in with cursor handling, since cursor handling is already complex. I've been thinking about what I could suggest instead, until we get the operator-based solution working. Here's an idea.

Leave the cursor layer as-is: it produces a set of rows as Java arrays. If the query runner sees that a sort is requested, insert another `Sequence` layer above the cursor. The native query stuff is a bit messy, but here's the gist.

Each cursor represents one slice of time. If the sort spec starts with `__time`, then the results from each cursor can be concatenated to produce the sorted result for a segment.

If the key isn't `__time`, then do a two-level sort-merge. Use the sorter from above to sort the results from each cursor. Hold on to all these sorted lists until all cursors are done. Then, merge the results using the priority queue mechanism you have in the above file. Voila!, a sorted per-segment result. Results from multiple segments must be merged, but the code to check for merges already exists. You may have to modify it to handle your custom merge.

The one drawback of this, mentioned above, is memory pressure. But, since there is no good solution at present, we can start with ignoring the memory issue.

This layered trick is what we've done for operators; so your solution will have all the pieces needed when we convert this code: we'll just wrap your solution in an operator interface rather than a `Sequence` interface.